### PR TITLE
Modify old differentiable optimizer tests to use optim_db

### PR DESCRIPTION
**Pull request by emmett-bicker**
Fourth PR in a larger project to broaden differentiable optimizer support with @janeyx99 ! This one is the first step in Step 0.

This PR replaces the 14 old differentiable tester functions with one differentiable tester function that uses optim_db and assures that the gradient can flow through the optimizer wrt params for all accepted OptimizerInput values attributed to each OptimizerInfo.

It also revealed a potentially confusing case where SGD defaults to foreach on cuda, and as it doesn't have a differentiable flag, it just throws a normal error that makes it look like a differentiable optimizer isn't supported, so I'll create an issue that brings that up!

Also, I added one typehint and fixed another in common_optimizers. This might not be great for organization so I can move it into another PR if desired. 

The tests used to take 0.7 seconds, but now that they run every OptimizerInput test, they now take around 7 seconds. Ever time  it runs gradcheck, it prints out a warning for using a deprecated version of vmap in gradcheck (which also happened last time) -- I believe fixing this warning would cut a lot of time off the tests since it seems largely consumed by printing!

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar @mruberry @ZainRizvi

### CODE REVIEW

**Comment by EmmettBicker** [comment]
Idk if I'm supposed to add the labels but I don't see any harm in me doing that rn!

File: torch/testing/_internal/common_optimizers.py [hunk 1]
```diff
@@ -104,12 +104,12 @@ class OptimizerInfo:
...
     def __init__(
         self,
-        optim_cls: Optimizer,  # Class object for the Optimizer under test
+        optim_cls: Type[Optimizer],  # Class object for the Optimizer under test
         *,
         # Function to generate optimizer inputs EXCLUDING params. We delegate params responsibility
         # to the test using the OptimizerInfo. OptimizerInput.params is likely None.
         # Can optionally take in device to filter out certain unsupported configs
-        optim_inputs_func,
+        optim_inputs_func: Callable[..., List[OptimizerInput]],
```
Comment (Skylion007) [inline torch/testing/_internal/common_optimizers.py:112]: Don't erase the args to the function, use a typing_extensions.ParamSpec variable.

Comment (EmmettBicker) [reply torch/testing/_internal/common_optimizers.py:112]: Ok! That sounds great

**Comment by EmmettBicker** [comment]
Actually. I'm going to move all these tests into the normal test_optim so please don't merge yet!

File: torch/testing/_internal/common_optimizers.py (hunk 1 above)
Comment (EmmettBicker) [reply torch/testing/_internal/common_optimizers.py]: They all seemed to have the same typing pattern so I used protocol, is that okay?

**Comment by EmmettBicker** [comment]
Im not sure if the line 

```py
# load_tests from common_utils is used to automatically filter tests for
# sharding on sandcastle. This line silences flake warnings
load_tests = load_tests
```

is still necessary

**Comment by EmmettBicker** [comment]
All these force pushes are bc #143726 got merged and I'm just making sure everything looks good

File: torch/testing/_internal/common_optimizers.py [hunk 2]
```diff
@@ -99,17 +99,24 @@ def __init__(
         self.error_regex = error_regex
 
 
+# Typing for every optim_inputs_func
+class OptimInputsFunc(Protocol):
```
Comment (Skylion007) [inline torch/testing/_internal/common_optimizers.py:103]: Why does this need to be a protocol? JIT support?

Comment (EmmettBicker) [reply torch/testing/_internal/common_optimizers.py:103]: It probably doesn't! I just thought that in this typing structure you're supposed to use a Protocol, but I might've learned incorrectly and have been unnecessarily adding Protocol

Comment (EmmettBicker) [reply torch/testing/_internal/common_optimizers.py]: Just got rid of it, thanks for pointing that out!

**Comment by janeyx99** [comment]
> Should I be adding in my own labels w/ pytorchbot? I'm thinking of adding 'not user facing.' I'm going to do it rn because I dont see the harm but I'd love to know if this is something I should usually be doing!

yes you can do this!

File: test/optim/test_optim.py
Comment (janeyx99) [inline test/optim/test_optim.py:1]: let's actually keep the differentiable tests in this file, we can rename the test file to test_differentiable_optim.py.

Moving code around makes git blame'ing more annoying in the future, and it's usually good to maintain modularity here when differentiable tests are still pretty different.

File: test/test_optim.py [hunk 3]
```diff
@@ -2216,9 +2218,90 @@ def test_non_empty_state(self, device, dtype, optim_info):
...
+            mock_optim_kwargs = optim_input.kwargs.copy()
+            optim_state = _get_optimizer_state(
+                optim_cls, mock_params, mock_optim_kwargs
+            )
 
+            optim_kwargs = optim_input.kwargs
+            optim_kwargs.update({"differentiable": True})
+
+            # Cannot be capturable and differentiable
+            if "capturable" in optim_kwargs:
```
Comment (janeyx99) [inline test/test_optim.py:2282]: I'd rather skip these tests instead of duplicating the run.

Also, capturable and differentiable shouldn't be mutually exclusive--which optims are these failing for?

File: test/test_optim.py [hunk 4]
```diff
@@ -2216,9 +2218,90 @@ def test_non_empty_state(self, device, dtype, optim_info):
...
+            state = states[0] if 0 in states else states
+            return state
+
+        optim_cls = optim_info.optim_cls
+        if "differentiable" not in optim_info.supported_impls:
+            self.skipTest(f"Differentiable {optim_cls.__name__} not supported")
+
+        for optim_input in optim_info.optim_inputs_func(device=device):
+            mock_params = torch.rand(10, requires_grad=True, device=device, dtype=dtype)
+            mock_optim_kwargs = optim_input.kwargs.copy()
```
Comment (janeyx99) [inline test/test_optim.py:2273]: what's the diff between copy and deepcopy? / why is the copy needed at all? Calling optim constructors shouldn't change the kwargs, do they?

File: test/test_optim.py [hunk 5]
```diff
@@ -2216,9 +2218,90 @@ def test_non_empty_state(self, device, dtype, optim_info):
...
+            optim_kwargs = optim_input.kwargs
+            optim_kwargs.update({"differentiable": True})
+
+            # Cannot be capturable and differentiable
+            if "capturable" in optim_kwargs:
+                optim_kwargs.update({"capturable": False})
+
+            # sgd defaults to foreach on cuda so must explicitly specify not foreach
+            if optim_cls is SGD and device.startswith("cuda"):
+                optim_kwargs.update({"foreach": False})
```
Comment (janeyx99) [inline test/test_optim.py:2287]: I'm guessing you'll rebase this PR once the flag is added to SGD?

File: test/test_optim.py [hunk 6]
```diff
@@ -2216,9 +2218,90 @@ def test_non_empty_state(self, device, dtype, optim_info):
...
+                    grad,
+                    optim_state,
+                    optim_cls,
+                    optim_kwargs,
+                    *optim_state.values(),
+                ),
+            )
+
+
+instantiate_device_type_tests(TestOptimRenewed, globals(), allow_mps=True)
```
Comment (janeyx99) [inline test/test_optim.py:2304]: Can you include a printout of what test cases run for this? We want to make sure we're strictly increasing coverage.

File: torch/testing/_internal/common_optimizers.py [hunk 7]
```diff
@@ -99,17 +99,24 @@ def __init__(
         self.error_regex = error_regex
 
 
+# Typing for every optim_inputs_func
+class OptimInputsFunc:
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:103]: There's the shorter one line way of defining function types in python like https://github.com/pytorch/pytorch/blob/efab8c433f257a4ab01fd07ca4b4a6aaf6ddc632/torch/optim/optimizer.py#L47

How does this `def __call__...` compare?

File: test/test_optim.py [hunk 8]
```diff
@@ -2216,9 +2218,90 @@ def test_non_empty_state(self, device, dtype, optim_info):
...
+                if isinstance(v, torch.Tensor) and v.requires_grad
+            )
 
-instantiate_device_type_tests(TestOptimRenewed, globals(), allow_mps=True)
+        def _get_optimizer_state(
+            optim_cls: type[Optimizer],
+            params: torch.Tensor,
+            optim_kwargs: dict[str, Any],
+        ) -> dict[str, Any]:
+            params.grad = torch.rand_like(params)
```
Comment (janeyx99) [inline test/test_optim.py:2257]: params -> param for convention

**Review by janeyx99 — COMMENTED** [review]
love to see the code condense, left a first round review!
